### PR TITLE
Fix AWS iam_user remove

### LINKED ISSUE #59078: iam_user module fails to remove an AWS user
##### SUMMARY

`iam_user` module fails to remove am AWS user if it has: a group membership, console password (login profile), access keys, MFA device, etc..

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

iam_user

##### ANSIBLE VERSION

```paste below
ansible 2.8.2
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/home/agustin/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = .../venv/lib/python3.5/site-packages/ansible
  executable location =.../venv/bin/ansible
  python version = 3.5.7 (default, Mar 30 2019, 01:05:18) [GCC 7.3.0]
```

##### CONFIGURATION

```paste below

```

##### STEPS TO REPRODUCE

Create an AWS user `test` with password or access keys

```yaml
- hosts: localhost
  tasks:
    - iam_user:
        name: test
        state: absent
```

##### EXPECTED RESULTS

User's state = absent

##### ACTUAL RESULTS

```paste below
TASK [iam_user] **************************************************************************************************************************************
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: botocore.errorfactory.DeleteConflictException: An error occurred (DeleteConflict) when calling the DeleteUser operation: Cannot delete entity, must delete login profile first.
fatal: [localhost]: FAILED! => {"changed": false, "error": {"code": "DeleteConflict", "message": "Cannot delete entity, must delete login profile first.", "type": "Sender"}, "msg": "Unable to delete user test: An error occurred (DeleteConflict) when calling the DeleteUser operation: Cannot delete entity, must delete login profile first.", "response_metadata": {"http_headers": {"content-length": "300", "content-type": "text/xml", "date": "Sun, 14 Jul 2019 18:34:25 GMT", "x-amzn-requestid": "01da0109-a666-11e9-adeb-83c56b62b76d"}, "http_status_code": 409, "request_id": "01da0109-a666-11e9-adeb-83c56b62b76d", "retry_attempts": 0}}

```

### PR DESCRIPTION
**Pull request by tinproject**
##### SUMMARY

Fixes #59078 

As Boto3 documentation tells (https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/iam.html#IAM.Client.delete_user) before delete an user it is needed to delete some user's associated resources, like login profile(console password), access keys, etc..

> 
    Password ( DeleteLoginProfile )
    Access keys ( DeleteAccessKey )
    Signing certificate ( DeleteSigningCertificate )
    SSH public key ( DeleteSSHPublicKey )
    Git credentials ( DeleteServiceSpecificCredential )
    Multi-factor authentication (MFA) device ( DeactivateMFADevice , DeleteVirtualMFADevice )
    Inline policies ( DeleteUserPolicy )
    Attached managed policies ( DetachUserPolicy )
    Group memberships ( RemoveUserFromGroup )

The current implementation of `iam_user` don't take most of this into account so user's removal fails.
This PR removes all the user's associated resources before deletes the user.

**Rationale of PR**:
- I refactor the code a bit to be more clear.
- I don't include every step in it's own try/except as I don't believe it adds value over the already present: 'user can't be removed' with included stack-trace.

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

iam_user

##### ADDITIONAL INFORMATION

To test this I've created manually the following IAM users and remove them with the `iam_user` module of this PR:

- test1 -> simple user with access keys
- test2 -> simple user with login profile (console password)
- test3 -> user with: access key, group membership, virtual MFA, attached policy, ssh public key, signing certificate, git https credentials.
- test4 -> user with: login profile, group membership, U2F MFA, inline policy, ssh public key, git https credentials.

### CODE REVIEW

File: lib/ansible/modules/cloud/amazon/iam_user.py
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/iam_user.py]: Might be worth moving the module to AnsibleAWSModule and using `fail_json_aws`

**Review by jillr — APPROVED** [review]
tested with a user that has keys, console login, and policies, lgtm

**Comment by tinproject** [comment]
Any more requested changes to this PR?
Ping @s-hertel @willthames

**Comment by willthames** [comment]
rebuild_merge

**Comment by tinproject** [comment]
Close & re-open for CI to pass.

@willthames @jillr @anotherjesse Sorry to insist, is there any remaining issue so this could be on 2.9?

**Comment by tremble** [comment]
@tinproject it might help if you added some integration tests.  A reasonably "simple" example can be found at https://github.com/ansible/ansible/pull/55821/commits/6df50c3c35f4fe30698b95d9c99f532b31ce5d50 

I know it's kinda annoying to be asked for them when there's limited examples.

@willthames has some helpful documentation http://willthames.github.io/2017/07/17/so-you-want-to-test-aws-modules-for-ansible.html

**Comment by tinproject** [comment]
I had think about create integration tests but the `iam_user` module could not create any of the things that make the user deletion fail (like MFA or security credentials). I've ended up using a playbook with command tasks calling the aws-cli directly, but as I do not knew if to have that dependency on aws-cli is a valid case for integration tests I did not include it on this PR. I could share it.

Also (this is why I insist), `iam_user` module could be considered **broken** on removing, because everything that makes an IAM user _usable_, like: _security credentials_, _console login_, or _group membership_, makes the user deletion fail.

**Comment by tinproject** [comment]
Rebased on current `devel` branch. Squash PEP8 commit.

Add basic integration tests for `iam_user` module:

* The test consist on a user creation, group creation given the user the group membership, and user removal. 
* The test should fail on current released Ansible, but pass with the code of this PR. (I've tested with a custom playbook, didn't have time to create the test environment)

File: test/integration/targets/iam_user/tasks/main.yml
Comment (tinproject) [inline test/integration/targets/iam_user/tasks/main.yml]: I base this test on `iam_group` ones. This is different on https://github.com/ansible/ansible/blob/devel/test/integration/targets/iam_group/tasks/main.yml#L27 so I believe either my test is broken or `iam_group` module output is broken.

Documentation of `iam_group` goes in line with the `iam_group` test version, but not with the current module code..

File: test/integration/targets/iam_user/tasks/main.yml
Comment (tremble) [inline test/integration/targets/iam_user/tasks/main.yml]: Personally, I have a preference for adding module_defaults to the block below 

```
- name: Test that a user can be created, attached to a group, and then removed
  module_defaults:
    group/aws:
      aws_access_key: "{{ aws_access_key }}"
      aws_secret_key: "{{ aws_secret_key }}"
      security_token: "{{ security_token }}"
      region: "{{ aws_region }}"
  block:
```

File: test/integration/targets/iam_user/tasks/main.yml
Comment (tremble) [inline test/integration/targets/iam_user/tasks/main.yml]: please use `security_token | default(omit)` some folks (myself included) run the tests using just an access/secret key pair rather than using temporary tokens.

File: test/integration/targets/iam_user/tasks/main.yml
Comment (tremble) [inline test/integration/targets/iam_user/tasks/main.yml]: It would be better here to define the username in a defaults section and use `'{{ resource_prefix }}-user'` as the username, it reduces the risk of one test run messing with another

File: test/integration/targets/iam_user/tasks/main.yml
Comment (tremble) [reply test/integration/targets/iam_user/tasks/main.yml]: Unfortunately some of the tests **are** broken (including iam_group) see #61243

**Comment by tremble** [comment]
Thanks for adding the tests, in a perfect world they'd be more thorough and test the MFA/Credential examples you gave, but for now I'd say they're sufficient, since they at least test the general operation of iam_user.

The general refactor looks sane to me, I'll have a quick run through the tests this evening (European time) and then add my +1 (since jillr seems to value it)

File: test/integration/targets/iam_user/tasks/main.yml
Comment (tremble) [inline test/integration/targets/iam_user/tasks/main.yml]: Would be good to also test behaviour with Policies (both managed and in-line) attached.  This can be done through Ansible, if you're not sure how I can dig out some examples later.

**Comment by tremble** [comment]
@tinproject Generally looks good although it'd be good if you'd pull in https://github.com/tremble/ansible/commit/5b766cfca170c5b6cdd5b74a205a7c931ffd3010 (more comprehensive tests)

@jillr code looks sane to me, and the results from my more comprehensive tests (see commit mentioned above) look good.  I'd be happy to submit my change over the top of @tinproject's.

**Comment by tremble** [comment]
lgtm

**Comment by tinproject** [comment]
Make requested changes to tests

**Comment by tremble** [comment]
Those changes certainly meet what I'd asked for
lgtm

**Comment by tremble** [comment]
@jillr I'll add some extra tests to make the iam_user integration tests more thorough once these land, but the integration tests already included are good for testing this bug.

File: lib/ansible/modules/cloud/amazon/iam_user.py
Comment (tremble) [inline lib/ansible/modules/cloud/amazon/iam_user.py]: Sorry for the late update, just spotted this (caught me out somewhere else)

[fail_json_aws is defined](https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/aws/core.py#L191) as

`fail_json_aws(self, exception, msg=None):`

You can simplify this to 

```
module.fail_json_aws(e, msg="Unable to delete user {0}".format(user_name))
```

Probably worth swapping out most of the fail_json calls for similar fail_json_aws calls too.

File: lib/ansible/modules/cloud/amazon/iam_user.py
Comment (tremble) [inline lib/ansible/modules/cloud/amazon/iam_user.py]: Bad fail_json_aws call here too

File: lib/ansible/modules/cloud/amazon/iam_user.py
Comment (tremble) [inline lib/ansible/modules/cloud/amazon/iam_user.py]: and another bad fail_json_aws call

File: test/integration/targets/iam_user/aliases
Comment (tremble) [inline test/integration/targets/iam_user/aliases]: @jillr did the User/Group pieces make it into Terminator yet? (is 'unsupported' needed here)

**Review by tremble — CHANGES_REQUESTED** [review]
Sorry for the slow/late niggles.

File: test/integration/targets/iam_user/aliases
Comment (jillr) [reply test/integration/targets/iam_user/aliases]: @tremble I don't see anything PRd yet.  For user management we'll also need to make sure there's sufficient guardrails to prevent that terminator and the tests from accessing the actual CI account users though so there will need to be a conversation around how best to do that.  I'd rather let this merge unsupported and enable the tests once we have the CI support in place than block on that.

Comment (tinproject) [reply test/integration/targets/iam_user/aliases]: Right now the `iam_user` module is **broken** and can not delete any _real_ AWS user.

It's being almost two months since I open this PR, merge or decline, please.

Comment (jillr) [reply test/integration/targets/iam_user/aliases]: We definitely do want this PR, which is why folks have been engaging on it and testing, reviewing, providing feedback. We want to ensure that things are in a good state and that we have tests that cover this module so that it won't break again in the future.  

There's policy and permissions things that we have to take into consideration for anything that will run automatically in the CI environment, especially when we're dealing with modules that can affect access, which is what tremble was asking about.  Sometimes we will block a change to get those pieces in place first but in this case I don't think we should do that.

Generally we like to give other reviewers who have been involved with a PR, like @tremble and @willthames, a chance to update their reviews to make sure the community has had an opportunity to review the final state of the code.  Ansible is a very big, busy project and a lot of our contributors, especially on the AWS modules, are community members like yourself who are also balancing other jobs or responsibilities.  We appreciate your patience if it takes us a bit to respond sometimes.

**Comment by tremble** [comment]
@tinproject needs a rebase because #23382 finally landed

File: lib/ansible/modules/cloud/amazon/iam_user.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/iam_user.py]: I see you didn't change this from before, but please update this to handle BotoCoreError instead here, as ParamValidationError is only a subset of the things that could go really wrong. You'll need to import it too.

Since you ported this to AnsibleAWSModule it's preferable to use fail_json_aws and handle these two exceptions together; ClientError is only handled separately if you need to parse the .response['Error']['Code'] because it doesn't exist for BotoCoreError. `fail_json_aws` cuts down on boilerplate (it will add the traceback, to_native(e), and e.response for you if these things exist and handle it gracefully if not).
```suggestion
    except (ClientError, BotoCoreError) as e:
        module.fail_json_aws(e, msg="Unable to detach policy {0} from user {1}".format(policy['PolicyArn'], user_name))
```

File: lib/ansible/modules/cloud/amazon/iam_user.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/iam_user.py]: This too can just be
```suggestion
    except (ClientError, BotoCoreError) as e:
        module.fail_json_aws(e, msg="Unable to delete user {0}".format(user_name))
```

**Review by s-hertel — COMMENTED** [review]
Couple suggestions to clean up the exception handling, but looks good to me once those are addressed.

**Review by tremble — CHANGES_REQUESTED** [review]
@tinproject it looks like the rebase has dropped the tests you'd put in place to test behaviour when deleting users which still have policies/group membership attached to them.

Please would you add them back.

**Comment by tinproject** [comment]
I made this PR to **FIX** the `iam_user` **deletion** because I was tired to do it manually every time there's the need to remove an user on my job. I was so naive to use this module to automate IAM because it **never worked** for real IAM users removal.

I've done this on my free time, and my allotted free time is well pass through. My original intention is to provide a fix to user deletion, nor to update the module to use `AnsibleAWSModule`, nor to add integration tests that not exist before and then conflict completely with others, with completely different style, of other module that use this module test file..

I believe Red Hat Ansible must rethink it's relationship with the community about what is needed on  PR and what are _nice to haves_. More than three months to add a fix, it have no sense.

@tremble Feel free to recover the tests. I'm moving my IAM user process to use Terraform/Cloudformation.

**Comment by s-hertel** [comment]
@tinproject I understand your frustration. There is only one Red Hatter working (mostly) full time on AWS pull requests and it's really a community-driven process. Most of the active community is in the IRC channel #ansible-aws on irc.freenode.net. It's a good place to ping people if you feel a pull request is stuck or feel that unreasonable demands have been made or just need some help implementing changes.

Updating the module to use AnsibleAWSModule was not required, but it did simplify the exception handling code. 

Integration tests are required for new modules and features, but depending on the urgency of the bugfix may also be excluded. Generally tests are requested for bugs and are considered more important than just a "nice to have" because they prevent the use case being broken in the future. As most of these modules are community maintained and often have new contributors, breakage happens easily without tests. You can see from the history of https://github.com/ansible/community/blob/master/group-aws/integration.md that we are progressing, though it's slow work.

It's often difficult to get the balance right between getting a PR in as-is, and forcing contributors to spend a lot of time improving it, refactoring other bits of code, and adding tests, we may have failed you in this occurrence, and for that we apologize.

Given the reviews so far, and the detailed manual testing you provided I think we are good to merge your PR as is.

**Comment by tinproject** [comment]
Thanks!

**Pull request merged**
